### PR TITLE
Merge bug/car-91 with Master

### LINKED ISSUE #91: Car drives into obstacle first time when booted
### Description
As a driver, I want the car's behavior to be the same whether it is the first time or the nth time it is executed after booting the car, so that I can have a good idea of what the car is gonna do if I trigger a specific action.

### Acceptance criteria
The car doesn't drive into the obstacle the first time it is booted.

### PR DESCRIPTION
**Pull request by majawennroth**
This has not solved the issue with driving into the obstacle on the first try ONLY. Start convo with Platis.

### CODE REVIEW

**Review by oscarhjern — COMMENTED** [review]
Line 107/108: 
The sensor return the correct value in centimeters (we have tested it). When the value is lower than STOP_DIST (15 cm) does the car stop. This happens all the time, without fail. We, therefore, conclude that the front sensor works as intended and that the logic for stopping (when there is an obstacle in front of the car) works as intended.

File: car/car.ino [hunk 1]
```diff
@@ -143,8 +144,8 @@ void handleInput() { //handle serial input (String!!)
 void obstacleAvoidance(){
   int leftDistance = left.readRangeContinuousMillimeters();
   int rightDistance = right.getMedianDistance();
-
-  if (leftDistance > LEFT_DIST && rightDistance > RIGHT_DIST){   //if there is no obstical on either sides rotate 90 degrees to the left
+  
+  if (leftDistance > LEFT_DIST && rightDistance > RIGHT_DIST || leftDistance > LEFT_DIST && rightDistance == 0){   //if there is no obstical on either sides rotate 90 degrees to the left
```
Comment (majawennroth) [inline car/car.ino:148]: @platisd When the car enter the if statment (or the else if/else) the first time after turning on the car. it continues to drive into the obstacle that has mede it stop in the first place. meaning that the car stops when i has detected an obstacle but then continues to drive into the obstical. once the car has stoped and you test it again (without turning it off) stop in front of the obstical (without hitting it) and then drives around it as it should. (Like in the demo we showed you) We have printed the values from the sensors while testing and nothing unexpected have appeard.

do you have any ideas of what could be wrong?

File: car/car.ino [hunk 2]
```diff
@@ -143,8 +144,8 @@ void handleInput() { //handle serial input (String!!)
 void obstacleAvoidance(){
   int leftDistance = left.readRangeContinuousMillimeters();
```
Comment (oscarhjern) [inline car/car.ino:145]: Row 145-147:
rightDistance and leftDistance get value from the sensors. We tested and printed the values. 

- leftDistance was always >1000 (mm) 
- rightDistance was always == 0 (cm) 

That means that leftDistance was always larger than LEFT_DIST (500 mm) and rightDistance was always == 0 following the logigc of the if-statment. 

We also tested the if-statements by printing letters for each. They work as intended meaning if there is no object on either side of the statement it goes into the first if, if on the left side it goes into the second if, if on the right side it goes into the third and if objects on all sides it goes into the else. 

We have located that the problem should be in the if-statement, somewhere, but we can not figure out what might couse the problem.

File: car/car.ino (hunk 1 above)
Comment (platisd) [reply car/car.ino:148]: Responding to both @majawennroth @oscarhjern 

I can't 100% follow, but have you verified that `rotateOnSpot` works as intended when this behavior is exhibited?

The logic of the `if` statements doesn't seem wrong which agrees with the results of your investigation. On a different and unrelated note, I would suggest adding more parentheses in order to clarify which comparisons go together. ([ES.41](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-parens))
I had to look up [operator precedence](https://en.cppreference.com/w/cpp/language/operator_precedence) in C++ to make sure the expression (at [L148](https://github.com/DIT112-V20/group-07/blob/dd1d1df90f7cd77218feacb982495871ed707a60/car/car.ino#L148)) would be evaluated the way you intend.

Anyway, it sounds like for currently unknown reasons `rotateOnSpot` doesn't *actually* rotate the car by `TURN_LEFT` degrees. Printing out the degrees returned by the gyroscope via `car.getHeading()` (at [L76](https://github.com/DIT112-V20/group-07/blob/dd1d1df90f7cd77218feacb982495871ed707a60/car/Car_Control.ino#L76) and [L83](https://github.com/DIT112-V20/group-07/blob/dd1d1df90f7cd77218feacb982495871ed707a60/car/Car_Control.ino#L83)) will shed some more light on the situation.

https://github.com/DIT112-V20/group-07/blob/dd1d1df90f7cd77218feacb982495871ed707a60/car/Car_Control.ino#L76-L101

Comment (platisd) [reply car/car.ino:148]: Long shot but could be a quick fix: Add a `car.update()` in the `loop()`. As [explained in the documentation](https://platisd.github.io/smartcar_shield/class_smart_car.html#ae6cd4a31632721f3a808c36da15cd5d3) it is suggested that you have this being executed as often as possible. In your case you have it only when you want to rotate which makes me suspect that there may be a lot of error the first time you call `car.getHeading()`.

Comment (oscarhjern) [reply car/car.ino:148]: **Update:** Car.update() worked!! @platisd

**Review by johann-tam — APPROVED** [review]

**Pull request merged**